File: src/algorithms/cpp/Wiggle_Sort_II.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
typedef long long LL;
const LL maxll = 0xfffffffff;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}


class Solution {
private:
    int get_median(vector <int>& nums) {
        int k = (SZ(nums) - 1) / 2 + 1;
        int l = -214748364, r = 214748364;
        while (l < r) {
            int mid = (l < 0 && r > 0) ? (l + r) / 2 : l + (r - l) / 2;
            int larger_cnt = 0, small_cnt = 0, tag = 0;
            for (int &num: nums) {
                if (num < mid) {
                    if (++ small_cnt > k - 1) {
                        tag = 1;
                        break;
                    }
                }
                else if (num > mid) {
                    if (++ larger_cnt > (SZ(nums) - k)) {
                        tag = -1;
                        break;
                    }
                }
            }
            if (tag == 0)
                return mid;
            else if (tag < 0)
                l = mid + 1;
            else
                r = mid;
        }
        return -1;
    }

public:
    void wiggleSort(vector <int>& nums) {
        int median = get_median(nums);
        int mod = SZ(nums) | 1;
        int i = 1 % mod, j = 1 % mod, k = (2 * SZ(nums) - 1) % mod;
        for (int t = 0; t < SZ(nums); ++ t) {
            if (nums[j] > median) {
                swap(nums[i], nums[j]);
                i = (i + 2) % mod;
                j = (j + 2) % mod;
            }
            else if (nums[j] < median) {
                swap(nums[j], nums[k]);
                k = (k - 2 + mod) % mod;
            }
            else {
                j = (j + 2) % mod;
            }
        }
    }
};


int main() {
    Solution solution = Solution();
    vector <int> v = {1, 3, 2, 2, 3, 1};
    solution.wiggleSort(v);
    return 0;
}

File: src/algorithms/python/Wiggle_Sort_II.py
#!/bin/env python2
# -*- coding: utf-8 -*-


class Solution(object):

    def get_median(self, nums):
        l, r, k = -214748364, 214748364, (len(nums) - 1) / 2 + 1
        while l < r:
            mid = l + (r - l) / 2 if l < 0 and r > 0 else (l + r) / 2
            larger_cnt, smaller_cnt, tag = 0, 0, 0
            for num in nums:
                if num < mid:
                    smaller_cnt += 1
                    if smaller_cnt > k - 1:
                        tag = 1
                        break
                elif num > mid:
                    larger_cnt += 1
                    if larger_cnt > len(nums) - k:
                        tag = -1
                        break
            if tag == 0:
                return mid
            else:
                l, r = (mid + 1, r) if tag < 0 else (l, mid)
        return -1

    def wiggleSort(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        median, mod = self.get_median(nums), len(nums) | 1
        i, j, k = 1 % mod, 1 % mod, (2 * len(nums) - 1) % mod
        for t in range(len(nums)):
            if nums[j] > median:
                nums[i], nums[j] = nums[j], nums[i]
                i, j = (i + 2) % mod, (j + 2) % mod
            elif nums[j] < median:
                nums[j], nums[k] = nums[k], nums[j]
                k = (k - 2 + mod) % mod
            else:
                j = (j + 2) % mod
